src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2WebSocketTests.cs

```diff
@@ -0,0 +1,122 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Net.Http;
+using System.Net.Http.HPack;
+using System.Runtime.ExceptionServices;
+using System.Text;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Extensions.Logging;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests;
+
+// https://datatracker.ietf.org/doc/html/rfc8441
+public class Http2WebSocketTests : Http2TestBase
+{
+    [Fact]
+    public async Task HEADERS_Received_ExtendedCONNECTMethod_Accepted()
+    {
+        await InitializeConnectionAsync(async context =>
+        {
+            Assert.Equal(HttpMethods.Connect, context.Request.Method);
+            Assert.Equal("websocket", context.Features.Get<IHttpConnectFeature>().Protocol);
+            Assert.False(context.Request.Headers.TryGetValue(HeaderNames.Protocol, out var _));
+            Assert.Equal("http", context.Request.Scheme);
+            Assert.Equal("/chat", context.Request.Path.Value);
+            Assert.Equal("server.example.com", context.Request.Host.Value);
+            Assert.Equal("chat, superchat", context.Request.Headers.WebSocketSubProtocols);
+            Assert.Equal("permessage-deflate", context.Request.Headers.SecWebSocketExtensions);
+            Assert.Equal("13", context.Request.Headers.SecWebSocketVersion);
+            Assert.Equal("http://www.example.com", context.Request.Headers.Origin);
+
+            Assert.Equal(0, await context.Request.Body.ReadAsync(new byte[1]));
+        });
+
+        // HEADERS + END_HEADERS
+        // :method = CONNECT
+        // :protocol = websocket
+        // :scheme = https
+        // :path = /chat
+        // :authority = server.example.com
+        // sec-websocket-protocol = chat, superchat
+        // sec-websocket-extensions = permessage-deflate
+        // sec-websocket-version = 13
+        // origin = http://www.example.com
+        var headers = new[]
+        {
+            new KeyValuePair<string, string>(HeaderNames.Method, "CONNECT"),
+            new KeyValuePair<string, string>(HeaderNames.Protocol, "websocket"),
+            new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
```

Comment:
Should be https - from comment?

Reply:
Normally yes, but these in-memory tests bypass the TLS layer and Kestrel requires that the scheme match the transport.